File: packages/plugins/clipboard/Sources/TanoPluginClipboard/ClipboardPlugin.swift
import Foundation
import TanoBridge

#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

/// Native clipboard plugin for Tano.
///
/// Provides copy/read operations via the ``TanoPlugin`` protocol.
/// Uses `UIPasteboard` on iOS and `NSPasteboard` on macOS.
///
/// Supported methods: `copy`, `read`.
public final class ClipboardPlugin: TanoPlugin {

    // MARK: - TanoPlugin conformance

    public static let name = "clipboard"
    public static let permissions: [String] = []

    public init() {}

    // MARK: - Routing

    public func handle(method: String, params: [String: Any]) async throws -> Any? {
        switch method {
        case "copy":
            return try copyText(params: params)
        case "read":
            return readText()
        default:
            throw ClipboardPluginError.unknownMethod(method)
        }
    }

    // MARK: - copy

    private func copyText(params: [String: Any]) throws -> [String: Any] {
        let text = params["text"] as? String ?? ""
        #if canImport(UIKit)
        UIPasteboard.general.string = text
        #elseif canImport(AppKit)
        let pasteboard = NSPasteboard.general
        pasteboard.clearContents()
        pasteboard.setString(text, forType: .string)
        #endif
        return ["success": true]
    }

    // MARK: - read

    private func readText() -> [String: Any] {
        let text: String?
        #if canImport(UIKit)
        text = UIPasteboard.general.string
        #elseif canImport(AppKit)
        text = NSPasteboard.general.string(forType: .string)
        #else
        text = nil
        #endif

        if let text = text {
            return ["text": text]
        } else {
            return ["text": NSNull()]
        }
    }
}

// MARK: - Errors

public enum ClipboardPluginError: Error, LocalizedError {
    case unknownMethod(String)

    public var errorDescription: String? {
        switch self {
        case .unknownMethod(let m):
            return "Unknown clipboard plugin method: \(m)"
        }
    }
}

File: packages/android/src/main/kotlin/dev/tano/plugins/ClipboardPlugin.kt
package dev.tano.plugins

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import dev.tano.bridge.TanoPlugin

/**
 * Native clipboard plugin for Tano on Android.
 *
 * Provides copy/read operations via the [TanoPlugin] interface.
 * Uses [android.content.ClipboardManager] for clipboard access.
 *
 * Supported methods: `copy`, `read`.
 *
 * Mirrors the iOS ClipboardPlugin.
 *
 * @param context Android Context needed to access the ClipboardManager system service.
 */
class ClipboardPlugin(private val context: Context) : TanoPlugin {

    override val name: String = "clipboard"
    override val permissions: List<String> = emptyList()

    private val clipboardManager: ClipboardManager
        get() = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

    // -- Routing --

    override suspend fun handle(method: String, params: Map<String, Any?>): Any? {
        return when (method) {
            "copy" -> copyText(params)
            "read" -> readText()
            else -> throw IllegalArgumentException("Unknown clipboard plugin method: $method")
        }
    }

    // -- copy --

    private fun copyText(params: Map<String, Any?>): Map<String, Any?> {
        val text = params["text"] as? String ?: ""
        val clip = ClipData.newPlainText("Tano", text)
        clipboardManager.setPrimaryClip(clip)
        return mapOf("success" to true)
    }

    // -- read --

    private fun readText(): Map<String, Any?> {
        val clip = clipboardManager.primaryClip
        val text = if (clip != null && clip.itemCount > 0) {
            clip.getItemAt(0).text?.toString()
        } else {
            null
        }

        return mapOf("text" to text)
    }
}
